onecodex/models/collection.py

```diff
@@ -389,6 +391,171 @@ def taxonomy(self):
+        annotation : {'pathways', 'metacyc', 'eggnog', 'go', 'ko', 'ec', 'pfam', 'reaction'}, optional
+            Annotation data to return
+        taxa_stratified : 'bool', optional
+            Return taxonomically stratified data
+        metric : {'coverage', 'abundance'} for annotation=='pathways', {'rpk', 'cpm'} for other annotations
+            Metric values to return
+        fill_missing : 'bool', optional
+            Fill np.nan values
+        filler : 'float', optional
+            Value with which to fill np.nans
+        """
+        # validate args
+        valid_annotation_values = [
+            "pathways",
+            "metacyc",
+            "eggnog",
+            "go",
+            "ko",
+            "ec",
+            "pfam",
+            "reaction",
+        ]
+        if annotation not in valid_annotation_values:
+            raise ValueError(f"'annotation' must be one of {', '.join(valid_annotation_values)}")
+        if annotation == "pathways":
+            if metric not in ["coverage", "abundance"]:
+                raise ValueError(
+                    "if using annotation='pathways', 'value' must be one of ['coverage', 'abundance']"
+                )
+        elif metric not in ["cpm", "rpk"]:
+            raise ValueError(
+                f"if using annotation={annotation}, 'value' must be one of ['cpm', 'rpk']"
+            )
+
+        data = {}
+        all_features = set()
+        # iterate over functional profiles, subset data, and store in data dict
+        for profile in self._functional_profiles:
+            # get table from mainline
+            table = profile.table(annotation=annotation, taxa_stratified=taxa_stratified)
+            # filter by indicated metric
+            if not taxa_stratified:
+                table = table[table["metric"] == "total_" + metric]
+            else:
+                table = table[table["metric"] == metric]
+            # if taxa stratified, concatenate id and taxon_name
+            if taxa_stratified:
+                table["id"] = table["id"] + "_" + table["taxon_name"]
+            # store tables for later retrieval
+            data[profile.id] = dict(zip(table.id, table.value))
+            all_features.update(set(table["id"]))
+
+        features_to_ix = {}
+        feature_list = []
+        for ix, feature in enumerate(all_features):
+            features_to_ix[feature] = ix
+            feature_list.append(feature)
+
+        # initialize an array and fill it
+        array = np.full(shape=(len(data), len(features_to_ix)), fill_value=np.nan)
+        profile_ids = []
+        for profile_index, profile_id in enumerate(data):
+            for feature_id in data[profile_id]:
+                array[profile_index, features_to_ix[feature_id]] = data[profile_id][feature_id]
+            profile_ids.append(profile_id)
+        df = pd.DataFrame(
+            array, index=pd.Index(profile_ids, name="functional_profile_uuid"), columns=feature_list
```

Comment:
usually in the client library we refer to UUIDs as "IDs", so `"functional_profile_id"` would be more consistent with other parts of the library

Reply:
fixed